lib/msprime.c

```diff
@@ -3010,6 +3010,11 @@ msp_run_single_sweep(msp_t *self)
      * depending on the value of curr_step, and hopefully reintroduce the max_time
      * and max_steps parameters. */
 
+    ret = model->params.single_sweep.generate_trajectory(
+            &model->params.single_sweep, self, &num_steps, &time, &allele_frequency);
+    if (ret != 0) {
```

Comment:
The idea here is that we decouple the trajectory simulation here by having a function pointer, which is pass 
- a reference to the ``single_sweep`` object (and should therefore know the trajectory parameters; more on this in a bit)
- a reference to the simulator so that (a) it can ask it to compute population sizes at a given time; and (b) can use its RNG instance
- And pointers to num_steps, time and allele_frequency. The generate_trajectory function will allocate and fill these out, and the calling function is responsible for freeing them.

Reply:
great. using the function pointer makes all the sense in the world.